File: Tree/95UniqueBinarySearchTreesII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> generateTrees(int n) {
        if(n == 0) return {};
        return helper(1, n + 1);
    }
    vector<TreeNode*> helper(int start, int end){
        vector<TreeNode*> res;
        if(start == end) return {nullptr};
        
        for(int i = start; i < end; i++){
            vector<TreeNode*> l = helper(start, i);
            vector<TreeNode*> r = helper(i + 1, end);

            for(auto ll:l)
                for(auto rr:r){
                    TreeNode *node = new TreeNode(i);
                    node->left = ll;
                    node->right = rr;
                    res.push_back(node);
                }
        }
        
        return res;
        
    }
};

File: Tree/95UniqueBinarySearchTreesII.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def generateTrees(self, n):
        """
        :type n: int
        :rtype: List[TreeNode]
        """
        if n == 0:
            return []
        return self.helper(1, n + 1)
    
    def helper(self, start, end):
        if start == end:
            return None
        res = []
        for i in xrange(start, end):
            for l in self.helper(start, i) or [None]:
                for r in self.helper(i + 1, end) or [None]:
                    node = TreeNode(i)
                    node.left, node.right = l, r
                    res.append(node)
        return res